servicetalk-concurrent-api/src/testFixtures/java/io/servicetalk/concurrent/api/TestCollectingPublisherSubscriber.java

```diff
@@ -0,0 +1,213 @@
+     * {@link #onError(Throwable)}.
+     *
+     * @param verifyOnNextConsumed {@code true} to verify that all {@link #onNext(Object)} signals have been consumed
+     * and throw if not. {@code false} to ignore if {@link #onNext(Object)} signals have been consumed or not.
+     * @return the exception received by {@link #onError(Throwable)}.
+     * @throws InterruptedException If an interrupt occurs while blocking for the terminal event.
+     */
+    public Throwable awaitOnError(boolean verifyOnNextConsumed) throws InterruptedException {
+        onTerminalLatch.await();
+        assert onTerminal != null;
+        if (onTerminal == ON_COMPLETE) {
+            throw new IllegalStateException("wanted onError but Subscriber terminated with onComplete");
+        }
+        if (verifyOnNextConsumed) {
+            verifyAllOnNextProcessed();
+        }
+        return onTerminal;
+    }
+
+    /**
+     * Block until a terminal signal is received, throws if {@link #onError(Throwable)} and returns normally if
+     * {@link #onComplete()}. This method will verify that all {@link #onNext(Object)} signals have been consumed.
+     *
+     * @throws InterruptedException If an interrupt occurs while blocking for the terminal event.
+     */
+    public void awaitOnComplete() throws InterruptedException {
+        awaitOnComplete(true);
+    }
+
+    /**
+     * Block until a terminal signal is received, throws if {@link #onError(Throwable)} and returns normally if
+     * {@link #onComplete()}.
+     *
+     * @param verifyOnNextConsumed {@code true} to verify that all {@link #onNext(Object)} signals have been consumed
+     * and throw if not. {@code false} to ignore if {@link #onNext(Object)} signals have been consumed or not.
+     * @throws InterruptedException If an interrupt occurs while blocking for the terminal event.
+     */
+    public void awaitOnComplete(boolean verifyOnNextConsumed) throws InterruptedException {
+        onTerminalLatch.await();
+        assert onTerminal != null;
+        if (onTerminal != ON_COMPLETE) {
+            throw new IllegalStateException("wanted onComplete but Subscriber terminated with onError", onTerminal);
+        }
+        if (verifyOnNextConsumed) {
+            verifyAllOnNextProcessed();
+        }
+    }
+
+    /**
+     * Block for a terminal event.
+     *
+     * @param timeout The duration of time to wait.
+     * @param unit The unit of time to apply to the duration.
+     * @return {@code true} if a terminal event has been received before the timeout duration.
+     * @throws InterruptedException If an interrupt occurs while blocking for the terminal event.
+     */
+    public boolean pollTerminal(long timeout, TimeUnit unit) throws InterruptedException {
+        return onTerminalLatch.await(timeout, unit);
+    }
+
+    private void verifyAllOnNextProcessed() {
+        if (!items.isEmpty()) {
```

Comment:
You can use `pollAllOnNext()` to simplify:

```java
List<T> items = pollAllOnNext();
if (!items.isEmpty()) {
    throw new IllegalStateException(items.size() + " onNext items were not processed: " + items);
}
```

Reply:
I still have to iterate to build the string, so I'll just keep this for now.

Reply:
I think the above code I provided would output the string pretty close to what is built by the iteration but it is not a big deal, I was just intending to reduce code.